src/pow_fpga_accel.c

```
@@ -0,0 +1,161 @@
+static uint32_t *cpow_map;
+static int result;
+static int8_t otrytes[TRANSACTION_LEN];
+
+int pow_fpga_accel_init()
+{
+    ctrl_fd = 0;
+    in_fd = 0;
+    out_fd = 0;
+    devmem_fd = 0;
+    fpga_regs_map = 0;
+    cpow_map = 0;
+
+    ctrl_fd = fopen("/dev/cpow-ctrl", "r+");
+
+    if (ctrl_fd == NULL) {
+        perror("cpow-ctrl open fail");
+        exit(EXIT_FAILURE);
+    }
+
+    in_fd = fopen("/dev/cpow-idata", "wb");
+
+    if (in_fd == NULL) {
+        perror("cpow-idata open fail");
+        fclose(ctrl_fd);
+        exit(EXIT_FAILURE);
+    }
+
+    out_fd = fopen("/dev/cpow-odata", "rb");
+
+    if (out_fd == NULL) {
+        perror("cpow-odata open fail");
+        fclose(ctrl_fd);
+        fclose(in_fd);
+        exit(EXIT_FAILURE);
+    }
+
+    devmem_fd = open("/dev/mem", O_RDWR | O_SYNC);
+
+    if (devmem_fd < 0) {
+        perror("devmem open");
+        fclose(ctrl_fd);
+        fclose(in_fd);
+        fclose(out_fd);
+        exit(EXIT_FAILURE);
+    }
+
+    fpga_regs_map =
+        (uint32_t *) mmap(NULL, HPS_TO_FPGA_SPAN, PROT_READ | PROT_WRITE,
+                          MAP_SHARED, devmem_fd, HPS_TO_FPGA_BASE);
+    cpow_map = (uint32_t *) (fpga_regs_map + CPOW_BASE);
+
+    if (fpga_regs_map == MAP_FAILED) {
+        perror("devmem mmap");
+        close(devmem_fd);
+        fclose(ctrl_fd);
+        fclose(in_fd);
+        fclose(out_fd);
+        exit(EXIT_FAILURE);
+    }
+
+    return 1;
+}
+
+void pow_fpga_accel_destroy()
+{
+    fclose(in_fd);
+    fclose(out_fd);
+    fclose(ctrl_fd);
+
+    result = munmap(fpga_regs_map, HPS_TO_FPGA_SPAN);
+
+    close(devmem_fd);
+
+    if (result < 0) {
+        perror("devmem munmap");
+        exit(EXIT_FAILURE);
+    }
+    exit(EXIT_SUCCESS);
+}
+
+int8_t *PowFPGAAccel(int8_t *itrytes, int mwm, int index)
+{
+    char nonce_trits[NONCE_LEN];
+
+    size_t itrytelen = 0;
+    size_t itritlen = 0;
+
+    itrytelen = strnlen((char *) itrytes, TRANSACTION_LEN);
```

I think the length of transaction trytes should be fixed, which is always 2673 in IOTA.

---

Yes, you say that I agree, but it can be used to check whether length of  transaction trytes is correct, e.g. https://github.com/LampaLab/iota_fpga/blob/master/pow_accel_soc/software/curl_pow_hard/curl_pow_hard.c#L82

---

If the length of `itrytes` is not equal to `TRANSACTION_LEN`, it should return an error instead of continuing to do PoW.
I think every implementation should also add this check like this.
```
if (strlen((char *) itrytes) != TRANSACTION_LEN) return NULL;
```

---

This check is added in dcurl_entry() before doing PoW. Would it be better?